### ISSUE TITLE
Typescript external support

### ISSUE THREAD

**Issue by nicolas-zozol** [masala/masala-parser#88]
Writing a definition file embedded in the package, implementing the @types system: https://blog.angular-university.io/typescript-2-type-system-how-do-type-definitions-work-in-npm-when-to-use-types-and-why-what-are-compiler-opt-in-types/

**Comment by nicolas-zozol** [comment]
ParsecJ is a good source of inspiration for type definitions: https://github.com/jon-hanson/parsecj

**Comment by nicolas-zozol** [comment]
package.json must reference the typings: 

```json
"typings": "./masala.d.ts",
````
The problem is that typings is obsolete, so it could move quickly

**Comment by nicolas-zozol** [comment]
Found it: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

In package.json: 

```json
{
    "name": "@masala/parser",
    "main": "./src/build/index.js",
    "types": "./masala.d.ts"
}
```

**Comment by nicolas-zozol** [comment]
Good news, we can make solid overloading in type definitions with Typescript:

```ts
function reverse(string: string): string;
function reverse<T>(array: T[]): T[];
```
https://blog.mariusschulz.com/2016/08/18/function-overloads-in-typescript

**Comment by bd82** [comment]
Hello.

If you plan on developing the d.ts definitions APIs as part of this repository.
But keep this repository as a vanilla JS project.

There is a way to do so, while **also** ensuring that the APIs **match** the implemntation
using new TypeScript compiler options.

See: https://github.com/bd82/typescript_for_public_apis
For details on this suggested pattern.

**Comment by nicolas-zozol** [comment]
Yes, it will be a 100% pure JS lib, with a .d.ts available with `@types`
